runtime/sema/checker.go

```diff
@@ -220,13 +220,14 @@ func (checker *Checker) PopPurityScope() PurityCheckScope {
 	return scope
 }
 
+func (checker *Checker) EnforcePurity(operation ast.Element, purity FunctionPurity) {
+	if purity == FunctionPurityImpure {
+		checker.ObserveImpureOperation(operation)
+	}
+}
+
 func (checker *Checker) ObserveImpureOperation(operation ast.Element) {
 	scope := checker.CurrentPurityScope()
-	// purity is monotonic, if we already know this scope is impure, there's no need to continue
-	if scope.CurrentPurity != FunctionPurityView {
-		return
-	}
-	scope.CurrentPurity = FunctionPurityImpure
```

Comment:
Just curious, why was this removed?

Reply:
Tracking the current purity of the scope is a relic of the original inference behavior; the purity of the scope we are in can no longer change so we don't need to update a variable here.